File: app/src/main/java/ru/p3tr0vich/fuel/SyncProviderDatabase.java
package ru.p3tr0vich.fuel;

import android.content.ContentProviderClient;
import android.content.ContentProviderOperation;
import android.content.ContentUris;
import android.content.ContentValues;
import android.content.OperationApplicationException;
import android.database.Cursor;
import android.nfc.FormatException;
import android.os.RemoteException;
import android.support.annotation.NonNull;
import android.support.v4.util.LongSparseArray;
import android.text.TextUtils;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

import ru.p3tr0vich.fuel.helpers.ContentProviderHelper;
import ru.p3tr0vich.fuel.helpers.DatabaseHelper;
import ru.p3tr0vich.fuel.models.FuelingRecord;
import ru.p3tr0vich.fuel.utils.UtilsLog;

class SyncProviderDatabase {

    private static final String TAG = "SyncProviderDatabase";

    private static final String INSERT = "+";
    private static final String DELETE = "-";
    private static final String CLEAR = "~";

    private static final String SEPARATOR = "\t";

    private final ContentProviderClient mProvider;

    SyncProviderDatabase(ContentProviderClient provider) {
        mProvider = provider;
    }

    @NonNull
    public List<String> getSyncRecords(boolean getAllRecords, boolean addDeleteAll) throws RemoteException {
        List<String> result = new ArrayList<>();

        Cursor cursor = mProvider.query(getAllRecords ?
                ContentProviderHelper.Companion.getURI_DATABASE_SYNC_ALL() :
                ContentProviderHelper.Companion.getURI_DATABASE_SYNC_CHANGED(), null, null, null, null);

        if (addDeleteAll) result.add(CLEAR);

        if (cursor != null) {
            FuelingRecord fuelingRecord;

            if (cursor.moveToFirst()) {
                do {
                    fuelingRecord = DatabaseHelper.getFuelingRecordForSync(cursor);

                    result.add(DatabaseHelper.getBoolean(cursor, DatabaseHelper.TableFueling.Columns.DELETED_INDEX) ?
                            DELETE + SEPARATOR +
                                    Long.toString(fuelingRecord.getId()) :
                            INSERT + SEPARATOR +
                                    Long.toString(fuelingRecord.getId()) + SEPARATOR +
                                    Long.toString(fuelingRecord.getDateTime()) + SEPARATOR +
                                    Float.toString(fuelingRecord.getCost()) + SEPARATOR +
                                    Float.toString(fuelingRecord.getVolume()) + SEPARATOR +
                                    Float.toString(fuelingRecord.getTotal()));
                } while (cursor.moveToNext());
            }
            cursor.close();
        }

        return result;
    }

    public void syncDeletedRecords() throws RemoteException {
        // Полностью удалить записи, отмеченные как удалённые
        mProvider.delete(ContentProviderHelper.Companion.getURI_DATABASE_SYNC(), null, null);
    }

    public void syncChangedRecords() throws RemoteException {
        // Изменить записи, помеченные как изменённые, как не изменённые
        mProvider.update(ContentProviderHelper.Companion.getURI_DATABASE_SYNC(), new ContentValues(0), null, null);
    }

    public void updateDatabase(@NonNull List<String> syncRecords)
            throws FormatException, RemoteException {

        int size = syncRecords.size();

        UtilsLog.d(TAG, "updateDatabase", "syncRecords.size() == " + size);

        if (size == 0) return;

        long id;

        boolean clearDatabase = false;

        String[] stringValues;

        ContentValues values;

        LongSparseArray<ContentValues> records = new LongSparseArray<>();

        for (String syncRecord : syncRecords) {
            if (TextUtils.isEmpty(syncRecord)) continue;

            stringValues = TextUtils.split(syncRecord, SEPARATOR);

//            for (int i = 0; i < stringValues.length; i++)
//                UtilsLog.d(TAG, "updateDatabase", "stringValues[" + i + "] == " + stringValues[i]);

            if (CLEAR.equals(stringValues[0]))
                clearDatabase = true;
            else {
                try {
                    id = Long.valueOf(stringValues[1]);
                } catch (Exception e) {
                    throw new FormatException(TAG +
                            " -- updateDatabase: error id, syncRecord == " + Arrays.toString(stringValues));
                }

                switch (stringValues[0]) {
                    case INSERT:
                        records.put(id,
                                DatabaseHelper.getValuesForSync(
                                        id,
                                        Long.valueOf(stringValues[2]),
                                        Float.valueOf(stringValues[3]),
                                        Float.valueOf(stringValues[4]),
                                        Float.valueOf(stringValues[5])));

                        break;
                    case DELETE:
                        records.put(id, null);

                        break;
                    default:
                        throw new FormatException(TAG +
                                " -- updateDatabase: error stringValues[0] == " + stringValues[0] +
                                ", syncRecord == " + Arrays.toString(stringValues));
                }
            }
        }

        ContentProviderOperation operation;
        ArrayList<ContentProviderOperation> operations = new ArrayList<>();

        if (clearDatabase)
            operations.add(ContentProviderOperation
                    .newDelete(ContentProviderHelper.Companion.getURI_DATABASE())
                    .build());

        size = records.size();

        for (int i = 0; i < size; i++) {
            values = records.valueAt(i);

            if (values != null)
                operation = ContentProviderOperation
                        .newInsert(ContentProviderHelper.Companion.getURI_DATABASE())
                        .withValues(values)
                        .build();
            else
                operation = ContentProviderOperation
                        .newDelete(ContentUris.withAppendedId(ContentProviderHelper.Companion.getURI_DATABASE(),
                                records.keyAt(i)))
                        .build();

            operations.add(operation);
        }

        size = operations.size();

        UtilsLog.d(TAG, "updateDatabase", "operations.size() == " + size);

        if (size == 0) return;

        try {
            mProvider.applyBatch(operations);
        } catch (OperationApplicationException e) {
            throw new FormatException(TAG + " -- updateDatabase: applyBatch exception == " + e.toString());
        }
    }
}

File: app/src/main/kotlin/ru/p3tr0vich/fuel/sync/SyncProviderDatabase.kt
package ru.p3tr0vich.fuel.sync

import android.content.*
import android.nfc.FormatException
import android.os.RemoteException
import android.text.TextUtils
import androidx.collection.LongSparseArray
import ru.p3tr0vich.fuel.helpers.ContentProviderHelper
import ru.p3tr0vich.fuel.helpers.DatabaseHelper
import ru.p3tr0vich.fuel.models.DatabaseModel
import ru.p3tr0vich.fuel.models.FuelingRecord
import ru.p3tr0vich.fuel.utils.UtilsLog
import java.util.*

internal class SyncProviderDatabase(private val provider: ContentProviderClient) {

    @Throws(RemoteException::class)
    fun getSyncRecords(getAllRecords: Boolean, addDeleteAll: Boolean): List<String> {
        val result = ArrayList<String>()

        val cursor = provider.query(
                if (getAllRecords) {
                    ContentProviderHelper.URI_DATABASE_SYNC_ALL
                } else {
                    ContentProviderHelper.URI_DATABASE_SYNC_CHANGED
                }, null, null, null, null)

        if (addDeleteAll) {
            result.add(CLEAR)
        }

        if (cursor != null) {
            var fuelingRecord: FuelingRecord

            if (cursor.moveToFirst()) {
                do {
                    fuelingRecord = DatabaseHelper.getFuelingRecordForSync(cursor)

                    result.add(
                            if (DatabaseHelper.getBoolean(cursor, DatabaseModel.TableFueling.Columns.DELETED_INDEX)) {
                                DELETE + SEPARATOR + fuelingRecord.id.toString()
                            } else {
                                INSERT + SEPARATOR +
                                        fuelingRecord.id.toString() + SEPARATOR +
                                        fuelingRecord.dateTime.toString() + SEPARATOR +
                                        fuelingRecord.cost.toString() + SEPARATOR +
                                        fuelingRecord.volume.toString() + SEPARATOR +
                                        fuelingRecord.total.toString()
                            })
                } while (cursor.moveToNext())
            }
            cursor.close()
        }

        return result
    }

    @Throws(RemoteException::class)
    fun syncDeletedRecords() {
        // Полностью удалить записи, отмеченные как удалённые
        provider.delete(ContentProviderHelper.URI_DATABASE_SYNC, null, null)
    }

    @Throws(RemoteException::class)
    fun syncChangedRecords() {
        // Изменить записи, помеченные как изменённые, как не изменённые
        provider.update(ContentProviderHelper.URI_DATABASE_SYNC, ContentValues(0), null, null)
    }

    @Throws(FormatException::class, RemoteException::class)
    fun updateDatabase(syncRecords: List<String>) {

        var size = syncRecords.size

        UtilsLog.d(TAG, "updateDatabase", "syncRecords.size() == $size")

        if (size == 0) return

        var id: Long

        var clearDatabase = false

        var stringValues: Array<String>

        var values: ContentValues?

        val records = LongSparseArray<ContentValues>()

        for (syncRecord in syncRecords) {
            if (TextUtils.isEmpty(syncRecord)) continue

            stringValues = TextUtils.split(syncRecord, SEPARATOR)

            //            for (int i = 0; i < stringValues.length; i++)
            //                UtilsLog.d(TAG, "updateDatabase", "stringValues[" + i + "] == " + stringValues[i]);

            if (CLEAR == stringValues[0]) {
                clearDatabase = true
            } else {
                try {
                    id = java.lang.Long.valueOf(stringValues[1])
                } catch (e: Exception) {
                    throw FormatException("$TAG -- updateDatabase: error id, syncRecord == ${stringValues.contentToString()}")
                }

                when (stringValues[0]) {
                    INSERT -> records.put(id,
                            DatabaseHelper.getValuesForSync(
                                    id,
                                    java.lang.Long.valueOf(stringValues[2]),
                                    java.lang.Float.valueOf(stringValues[3]),
                                    java.lang.Float.valueOf(stringValues[4]),
                                    java.lang.Float.valueOf(stringValues[5])))
                    DELETE -> records.put(id, null)
                    else -> throw FormatException("$TAG -- updateDatabase: error stringValues[0] == ${stringValues[0]}, syncRecord == ${stringValues.contentToString()}")
                }
            }
        }

        var operation: ContentProviderOperation
        val operations = ArrayList<ContentProviderOperation>()

        if (clearDatabase) {
            operations.add(ContentProviderOperation
                    .newDelete(ContentProviderHelper.URI_DATABASE)
                    .build())
        }

        size = records.size()

        for (i in 0 until size) {
            values = records.valueAt(i)

            operation = if (values != null) {
                ContentProviderOperation
                        .newInsert(ContentProviderHelper.URI_DATABASE)
                        .withValues(values)
                        .build()
            } else {
                ContentProviderOperation
                        .newDelete(ContentUris.withAppendedId(ContentProviderHelper.URI_DATABASE,
                                records.keyAt(i)))
                        .build()
            }

            operations.add(operation)
        }

        size = operations.size

        UtilsLog.d(TAG, "updateDatabase", "operations.size() == $size")

        if (size == 0) return

        try {
            provider.applyBatch(operations)
        } catch (e: OperationApplicationException) {
            throw FormatException("$TAG -- updateDatabase: applyBatch exception == $e")
        }

    }

    companion object {
        private const val TAG = "SyncProviderDatabase"

        private const val INSERT = "+"
        private const val DELETE = "-"
        private const val CLEAR = "~"

        private const val SEPARATOR = "\t"
    }
}